### PR TITLE
fix(pipelines): `CodeBuildStep.partialBuildSpec` not used, buildspec control for legacy API

### LINKED ISSUE #15169: (pipelines): need full control of BuildSpec
So set things like `runtime-versions`, `caching`, etc.

Would be good if we could merge artifacts parts of buildspecs...

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by hoegertn**
this fixes #15169

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by rix0rrr — DISMISSED** [review]

**Review by rix0rrr — APPROVED** [review]

File: packages/@aws-cdk/pipelines/test/compliance/synths.test.ts [hunk 1]
```diff
@@ -1043,7 +1085,7 @@ behavior('can provide custom BuildSpec that is merged with generated one', (suit
           },
           phases: {
             pre_build: {
-              commands: ['installCustom', 'install1', 'install2'],
+              commands: arrayWith('installCustom'),
```
Comment (hoegertn) [inline packages/@aws-cdk/pipelines/test/compliance/synths.test.ts:1094]: This does no longer validate the ordering...

File: packages/@aws-cdk/pipelines/lib/legacy/actions/publish-assets-action.ts [hunk 2]
```diff
@@ -127,6 +135,7 @@ export class PublishAssetsAction extends CoreConstruct implements codepipeline.I
         },
       },
     });
+    this.buildSpec = props.buildSpec ? codebuild.mergeBuildSpecs(props.buildSpec, buildSpec) : buildSpec;
```
Comment (otaviomacedo) [inline packages/@aws-cdk/pipelines/lib/legacy/actions/publish-assets-action.ts:138]: So the idea is that the user can augment the existing build spec, but not override anything. Is that it?

Comment (rix0rrr) [reply packages/@aws-cdk/pipelines/lib/legacy/actions/publish-assets-action.ts:138]: Not entirely. Some things may be overwritten, but in general they will be merged.

Comment (otaviomacedo) [reply packages/@aws-cdk/pipelines/lib/legacy/actions/publish-assets-action.ts:138]: Yes, that's true. Objects and arrays will be merged. But for primitives (let's say, `run-as`), the one generated by the action has precedence over the one provided by the user.

**Review by rix0rrr — APPROVED** [review]

**Comment by alexpulver** [comment]
@hoegertn it worked for me now, thanks!

**Pull request merged**